src/v/ssx/include/ssx/sharded_int.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+
+#include <seastar/core/smp.hh>
+
+#include <algorithm>
+#include <atomic>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace ssx {
+
+/// sharded_int is an atomic primitive that has a local copy of an integer
+/// on each shard that can be independently updated and provides a helper for
+/// gathering the maximum of the per-shard values. The local copies do not have
+/// the same value, they are updated independently.
+template<typename T>
+requires std::is_integral_v<T>
+class sharded_int {
+public:
+    sharded_int(T value)
+      : _state(ss::smp::count, value) {}
+    ~sharded_int() noexcept = default;
+
+    sharded_int(sharded_int&& other) noexcept = default;
+    sharded_int& operator=(sharded_int&&) noexcept = default;
+
+    sharded_int(sharded_int const&) = delete;
+    sharded_int& operator=(sharded_int const&) = delete;
+
+    /// return the maximum value across the shards
+    ss::future<T> max() const {
+        std::vector<std::atomic<T>> state_copy(ss::smp::count);
```

Comment:
Why are these atomic?

This function could be called `reduce` and take a reducer.

Reply:
It was to ensure the changes to `state_copy` are visible when we read it, but I think sending the result back through the return value of `submit_to` would have been an alternative. Please correct me if I'm wrong but as far as I understand it, without `atomic<T>` or returning a `future<T>` via `submit_to`, we have no guarantees about the changes inside the lambda being visible to the caller shard.

Anyways, that's more for my curiosity now as I've removed this method and rewrote the `last_time_ms` handling using `map_reduce`.